src/core/electrostatics_magnetostatics/elc.cpp

```diff
@@ -247,6 +247,16 @@ static void add_dipole_force(const ParticleRange &particles) {
   gblcblk[2] = 0; // sum q_i
 
   for (auto const &p : local_particles) {
+    /* check if charged particles are in valid region */
+    if ((p.p.q != 0) && ((p.r.p[2] > elc_params.h) || (p.r.p[2] < 0))) {
+      if (p.r.p[2] < 0)
+        runtimeErrorMsg() << "Particle " << p.p.identity << " entered ELC gap "
```

Comment:
We probably want the same check in `double dipole_energy()`, in which case this block can be moved to a free function, e.g. `inline void elc_gap_sanity_check(Particle const &p)`. We also need to add a test to make sure the `runtimeErrorMsg()` is properly converted to a Python exception in the Cython code, e.g. with
```python
    def test_elc_gap_exception(self):
        ... # set up an ELC system, or re-use one that has already been set up in another test
        self.system.constraints.clear() # remove walls, if any was used to minimize the system
        self.system.part[0].pos = [box_l, box_l, box_l + elc_gap - 1]
        with self.assertRaises(Exception):
            self.system.analysis.energy()
        with self.assertRaises(Exception):
            self.integrator.run(2)
```